src/crate/client/converter.py

```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8; -*-
+#
+# Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
+# license agreements.  See the NOTICE file distributed with this work for
+# additional information regarding copyright ownership.  Crate licenses
+# this file to you under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.  You may
+# obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# However, if you have executed another commercial license agreement
+# with Crate these terms will supersede the license and you may use the
+# software solely pursuant to the terms of the relevant commercial agreement.
+"""
+Machinery for converting CrateDB database types to native Python data types.
+
+https://crate.io/docs/crate/reference/en/latest/interfaces/http.html#column-types
+"""
+import ipaddress
+from copy import deepcopy
+from datetime import datetime
+from enum import Enum
+from typing import Any, Callable, Dict, Optional, Union, List
+
+InputVal = Any
+
+
+def _to_ipaddress(value: Optional[str]) -> Optional[Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]:
+    """
+    https://docs.python.org/3/library/ipaddress.html
+    """
+    if value is None:
+        return None
+    return ipaddress.ip_address(value)
+
+
+def _to_datetime(value: Optional[float]) -> Optional[datetime]:
+    """
+    https://docs.python.org/3/library/datetime.html
+    """
+    if value is None:
+        return None
+    return datetime.utcfromtimestamp(value / 1e3)
```

Comment:
Dear @mfussenegger, @matriv, and @seut,

regarding @mfussenegger's proposal to let the user optionally specify a time zone, and make the machinery return a timezone-aware datetime object, I see two options here.

Let's assume the user defines a custom timezone object like
```python
>>> import datetime
>>> tz_mst = datetime.timezone(datetime.timedelta(hours=7), name="MST")
```

and a timestamp in Epoch like
```python
# Fri, 16 Sep 2022 11:09:16 GMT
timestamp = 1663326556
```

On this spot, which variant would be semantically the right one? Note that [`datetime.utcfromtimestamp`](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp) does not accept the `tz` keyword argument, it will always return a naive datetime object with `tzinfo=None`.

1. We can either use [`datetime.fromtimestamp`](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp), which **does** accept the `tz` keyword argument.
   ```python
   >>> datetime.datetime.fromtimestamp(timestamp, tz=tz_mst)
   datetime.datetime(2022, 9, 16, 18, 9, 16, tzinfo=datetime.timezone(datetime.timedelta(seconds=25200), 'MST'))
   ```

2. Or we can replace the `tzinfo` on the naive datetime object returned by `utcfromtimestamp`, making it timezone-aware.
   ```python
   >>> datetime.datetime.utcfromtimestamp(timestamp).replace(tzinfo=tz_mst)
   datetime.datetime(2022, 9, 16, 11, 9, 16, tzinfo=datetime.timezone(datetime.timedelta(seconds=25200), 'MST'))
   ```

I think option 1. is the right choice. So the implementation would look like something along the lines of:
```python

def set_timezone(self, tz):
  if not isinstance(tz, datetime.timezone):
    raise TypeError(f"Timezone object has wrong type: {tz}")
  self.timezone = tz

def _to_datetime(self, value):
  """
  Convert CrateDB's `TIMESTAMP` column to the native Python `datetime` object.
  When a timezone is given, return a timezone-aware object.
  Otherwise, return a naive object which is meant to be the time in the UTC time zone.
  """
  if self.timezone is not None:
    return datetime.fromtimestamp(value / 1e3, tz=self.timezone)
  else:
    return datetime.utcfromtimestamp(value / 1e3)
```

What do you think?

With kind regards,
Andreas.

Reply:
Another implementation could look like this, it is **always** using `datetime.fromtimestamp(value / 1e3, tz=self.timezone)`, where `self.timezone` would be `datetime.timezone.utc` by default, i.e. it will always return timezone-aware datetime objects.

While definitively easier and more straight-forward, it could be less performant when **not** aiming to use timezone-aware datetime objects and stay in "naive" land **on purpose**.

```python
def __init__(self):
  # Return datetype objects in UTC by default.
  self.timezone = datetime.timezone.utc

def set_timezone(self, tz):
  """
  Specify a time zone if you want to have your datetime objects in a time zone other than UTC.
  """
  if not isinstance(tz, datetime.timezone):
    raise TypeError(f"Timezone object has wrong type: {tz}")
  self.timezone = tz

def _to_datetime(self, value):
  """
  Convert CrateDB's `TIMESTAMP` column to a timezone-aware native Python `datetime` object.
  When no timezone is given, UTC is used.
  """
  return datetime.fromtimestamp(value / 1e3, tz=self.timezone)
```

Reply:
> definitively easier and more straight-forward

When thinking about it once more, I think I would prefer the implementation variant outlined in my latest post. In this way, it is always consistent that timezone-aware datetime objects are returned from the data converter machinery.

> it could be less performant when not aiming to use timezone-aware datetime objects and stay in "naive" land on purpose.

I think when users are aiming for raw speed, they would completely opt-out of the data type converter machinery at all.

Reply:
Hi again. Talk is cheap. I've converged a few more thoughts about this into a proposed patch at #445, completely separating the topic of timezone-aware `datetime` handling from this one.